File: matiasmasso/Mat.NET/DT2/DTOCsv.vb
Public Class DTOCsv
    Property Rows As List(Of DTOCsvRow)
    Property Title As String
    Property Filename As String

    Public Sub New(Optional sTitle As String = "", Optional sFilename As String = "")
        _Rows = New List(Of DTOCsvRow)
        _Title = sTitle
        If sFilename = "" Then sFilename = sTitle
        If sFilename = "" Then sFilename = "M+O downloaded file.csv"
        If sFilename.EndsWith(".csv") Then
        Else
            sFilename += ".csv"
        End If
        _Filename = sFilename
    End Sub

    Shared Function Factory(src As String) As DTOCsv
        Dim vblfs As Integer = src.Count(Function(c As String) c = vbLf)
        Dim vbcrlfs As Integer = src.Count(Function(c As String) c = vbCrLf)
        Dim splitchar As String = IIf(vblfs > vbcrlfs, vbLf, vbCrLf)

        Dim lines() As String = src.Split(splitchar)
        Dim retval As New DTOCsv
        For Each sLine As String In lines
            Dim oRow As New DTOCsvRow
            oRow.Cells = New List(Of String)
            For Each sCell As String In sLine.Split(";")
                oRow.Cells.Add(sCell)
            Next
            retval.Rows.Add(oRow)
        Next
        Return retval
    End Function

    Public Function AddRow() As DTOCsvRow
        Dim oRow As New DTOCsvRow
        oRow.Cells = New List(Of String)
        _Rows.Add(oRow)
        Return oRow
    End Function

    Public Shadows Function ToString() As String
        Dim sb As New System.Text.StringBuilder
        For Each oRow As DTOCsvRow In _Rows
            Dim FirstField As Boolean = True
            For Each oField As String In oRow.Cells
                If FirstField Then
                    FirstField = False
                Else
                    sb.Append(";")
                End If
                If oField IsNot Nothing Then
                    Dim InsertQuotes As Boolean = oField.Contains(";")
                    If oField.Contains(vbCrLf) Then InsertQuotes = True
                    If oField.Contains(",") Then InsertQuotes = True

                    If InsertQuotes Then sb.Append(Chr(34))
                    sb.Append(oField)
                    If InsertQuotes Then sb.Append(Chr(34))
                End If
            Next
            sb.AppendLine()
        Next
        Dim retval As String = sb.ToString
        Return retval
    End Function

    Public Function ToByteArray() As Byte()
        Dim src As String = Me.ToString()
        Dim retval As Byte() = New Byte(src.Length * 2 - 1) {}
        System.Buffer.BlockCopy(src.ToCharArray(), 0, retval, 0, retval.Length)
        Return retval
    End Function
End Class

Public Class DTOCsvRow
    Property Cells As List(Of String)

    Public Sub New()
        _Cells = New List(Of String)
    End Sub

    Public Sub AddCell(sField As String)
        _Cells.Add(sField)
    End Sub

    Public Sub AddCell(oAmt As DTOAmt)
        If oAmt Is Nothing Then
            _Cells.Add("")
        Else
            _Cells.Add(oAmt.Eur)
        End If
    End Sub

End Class

File: matiasmasso/Mat.NET/DTO/DTOCsv.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOCsv
    {
        public List<DTOCsvRow> Rows { get; set; }
        public string Title { get; set; }
        public string Filename { get; set; }

        public DTOCsv(string sTitle = "", string sFilename = "")
        {
            Rows = new List<DTOCsvRow>();
            Title = sTitle;
            if (sFilename == "")
                sFilename = sTitle;
            if (sFilename == "")
                sFilename = "M+O downloaded file.csv";
            if (sFilename.EndsWith(".csv"))
            {
            }
            else
                sFilename += ".csv";
            Filename = sFilename;
        }

        public static DTOCsv Factory(string src)
        {
            List<string> lines = src.toLinesList();
            DTOCsv retval = new DTOCsv();
            foreach (string sLine in lines)
            {
                DTOCsvRow oRow = new DTOCsvRow();
                oRow.Cells = new List<string>();

                foreach (string sCell in sLine.Split(';'))
                    oRow.Cells.Add(sCell);
                retval.Rows.Add(oRow);
            }
            return retval;
        }

        public DTOCsvRow AddRow()
        {
            DTOCsvRow oRow = new DTOCsvRow();
            oRow.Cells = new List<string>();
            Rows.Add(oRow);
            return oRow;
        }

        public new string ToString()
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            var quotesNeeded = new[] { ";", ",", Environment.NewLine };
            var lines = new List<string>();
            foreach (DTOCsvRow oRow in Rows)
            {
                sb = new System.Text.StringBuilder();
                bool FirstField = true;
                foreach (string oField in oRow.Cells)
                {
                    if (FirstField)
                        FirstField = false;
                    else
                        sb.Append(";");

                    if (oField != null)
                    {
                        if (quotesNeeded.Any(oField.Contains))
                            sb.AppendFormat("'{0}'", oField);
                        else
                            sb.Append(oField);
                    }
                }
                lines.Add(sb.ToString());
            }
            string retval = String.Join(Environment.NewLine, lines.ToArray());
            return retval;
        }

        public byte[] toByteArray()
        {
            string src = this.ToString();
            byte[] retval = new byte[src.Length * 2 - 1 + 1];
            System.Buffer.BlockCopy(src.ToCharArray(), 0, retval, 0, retval.Length);
            return retval;
        }
    }

    public class DTOCsvRow
    {
        public List<string> Cells { get; set; }

        public DTOCsvRow()
        {
            Cells = new List<string>();
        }

        public void AddCell(string sField)
        {
            Cells.Add(sField);
        }

        public void AddCell(DTOAmt oAmt)
        {
            if (oAmt == null)
                Cells.Add("");
            else
                Cells.Add(oAmt.Eur.ToString());
        }
    }
}
